Crypto ops

jenswi-linaro opened the pull request:
Changes hash, mac, cipher and authenc cryptos to use an `ops` pointer in the context and thus ties the correct functions to a crypto context. This gives a fine grained selection of implementation of algorithms which makes it easier to integrate with crypto accelerators which are probed during boot.

The price is adding ~800 bytes code/data.

jenswi-linaro commented:
The checkpatch warning is a false positive.

jforissier commented on core/crypto/crypto.c:
```diff
@@ -8,43 +8,94 @@
+	struct crypto_hash_ctx *c __maybe_unused = NULL;
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_MD5)
```
Minor - how about moving the `#ifdef`s to `crypto_impl.h`? (the less conditionals in C files, the better).

jenswi-linaro replied:
I was considering the same, I'll fix.

jforissier commented on core/crypto/crypto.c:
```diff
@@ -139,13 +140,27 @@ TEE_Result crypto_cipher_update(void *ctx __unused, uint32_t algo __unused,
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	uint32_t class = TEE_ALG_GET_CLASS(algo);
+
+	if (class != TEE_OPERATION_CIPHER && class != TEE_OPERATION_MAC)
```
`TEE_OPERATION_AUTHENC` missing?

jenswi-linaro replied:
I'll add it.

jforissier commented on core/crypto/crypto.c:
```diff
@@ -162,43 +162,107 @@ TEE_Result crypto_cipher_get_block_size(uint32_t algo, size_t *size)
+	struct crypto_mac_ctx *c __maybe_unused = NULL;
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
```
Same suggestion here about moving conditionals to `crypto_impl.h`.

jenswi-linaro replied:
Yes, I'll take care of all those.

jforissier commented on core/lib/libtomcrypt/mac.c:
```diff
@@ -0,0 +1,228 @@
+ * Message Authentication Code functions
+ *****************************************************************************/
+
+#if defined(CFG_CRYPTO_HMAC)
```
We could save ifdefs by splitting into two files:
```
srcs-$(CFG_CRYPTO_HMAC) += hmac.c 
srcs-$(CFG_CRYPTO_OMAC) += omac.c
```
This is cleaner IMO but I think you prefer otherwise? ;-)
It does not matter much anyway.

jenswi-linaro replied:
Agree, I'll fix.

jforissier commented on core/lib/libtomcrypt/authenc.c:
```diff
@@ -0,0 +1,439 @@
+#define TEE_CCM_TAG_MAX_LENGTH		16
+#define TEE_GCM_TAG_MAX_LENGTH		16
+
+#if defined(CFG_CRYPTO_CCM)
```
Here also, I think two files would be cleaner (`ccm.c` and `gcm_from_cryptolib.c` to reflect the config names?). Not a strong feeling either.

jforissier commented on core/crypto/crypto.c:
```diff
@@ -98,49 +98,112 @@ TEE_Result crypto_hash_final(void *ctx, uint32_t algo __unused,
+	struct crypto_cipher_ctx *c __maybe_unused = NULL;
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_AES)
```
Conditionals in `crypto_impl.h` here also

jforissier commented on core/lib/libtomcrypt/cipher.c:
```diff
@@ -0,0 +1,623 @@
+	}
+}
+
+#if defined(CFG_CRYPTO_ECB)
```
...and split files? (that get_des2_key() is annoying, though, could it be inlined in a `.h`?)

jenswi-linaro commented:
Comments addressed tag applied
Original patches at https://github.com/jenswi-linaro/optee_os/tree/crypto_ops_orig
Updated patches with updates in separate commits at https://github.com/jenswi-linaro/optee_os/tree/crypto_ops2

jforissier commented on core/crypto/cbc-mac.c:
```diff
@@ -0,0 +1,220 @@
+
+TEE_Result crypto_des_cbc_mac_pkcs5_alloc_ctx(struct crypto_mac_ctx **ctx)
+{
+	return crypto_cbc_mac_alloc_ctx(ctx, TEE_ALG_DES_CBC_NOPAD, true);
```
Same.

jenswi-linaro commented:
Fixed the xtest error (CMAC vs OMAC confusion in config).  Renamed `omac.c` to `cmac.c` to make it more intuitive. Fixed the checkpatch error introduced.

jenswi-linaro commented on core/crypto/cbc-mac.c:
```diff
@@ -0,0 +1,220 @@
+
+TEE_Result crypto_des3_cbc_mac_pkcs5_alloc_ctx(struct crypto_mac_ctx **ctx)
+{
+	return crypto_cbc_mac_alloc_ctx(ctx, TEE_ALG_DES3_CBC_NOPAD, true);
```
No, this is correct. `TEE_ALG_DES3_CBC_MAC_PKCS5` is implemented using `TEE_ALG_DES3_CBC_NOPAD` and then some special dealing with padding.

jforissier replied:
OK makes sense.

jenswi-linaro commented on core/lib/libtomcrypt/src/tee_ltc_provider.c:
```diff
@@ -128,8 +128,7 @@ static void tee_ltc_reg_algs(void)
 
-#if defined(_CFG_CRYPTO_WITH_HASH) || defined(CFG_CRYPTO_RSA) || \
-	defined(CFG_CRYPTO_HMAC)
+#if defined(CFG_CRYPTO_RSA)
```
This function is from now on only used by code enabled with `CFG_CRYPTO_RSA`. I suppose it's possible to clean up and move the function, but this is with as little change as possible.

jforissier replied:
Hmmm. Indeed. Then, all the "case" statements that do not contain "RSA" are highly confusing. I think they should be removed.

jenswi-linaro replied:
Done, I noticed it's needed by `CFG_CRYPTO_DSA` too so I added that.

jforissier replied:
Wait!... How is `tee_algo_to_ltc_hashindex()` needed for DSA? I see it called from `crypto_acipher_rsaes_decrypt()`, `crypto_acipher_rsaes_encrypt()`, `crypto_acipher_rsassa_sign()` and `crypto_acipher_rsassa_verify()`. All deal with RSA only.

jenswi-linaro replied:
You're right, I'll remove the DSA stuff too.

jforissier commented on core/lib/libtomcrypt/src/tee_ltc_provider.c:
```diff
@@ -143,54 +143,41 @@ static TEE_Result tee_algo_to_ltc_hashindex(uint32_t algo, int *ltc_hashindex)
 	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA1:
 	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1:
-	case TEE_ALG_SHA1:
-	case TEE_ALG_DSA_SHA1:
```
Keep that one! (for CFG_CRYPTO_DSA)

jenswi-linaro replied:
Sorry, fixed

jenswi-linaro commented:
Update

jforissier commented on core/lib/libtomcrypt/src/tee_ltc_provider.c:
```diff
@@ -128,8 +126,7 @@ static void tee_ltc_reg_algs(void)
 
-#if defined(_CFG_CRYPTO_WITH_HASH) || defined(CFG_CRYPTO_RSA) || \
-	defined(CFG_CRYPTO_HMAC)
+#if defined(CFG_CRYPTO_RSA) || defined(CFG_CRYPTO_DSA)
```
I don't get it. IMO that should be `#if defined(CFG_CRYPTO_RSA)` and `case TEE_ALG_DSA_SHA1:` at line 146 should be removed.

jenswi-linaro replied:
I'm sorry, it seems I forgot to push the last patch.

jforissier replied:
No problem. Looks OK now.

jforissier commented:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

Please squash the fixups and we should be good to go. Thanks.

jenswi-linaro commented:
Done
